sys/kern/src/arch/riscv64.rs

```diff
@@ -0,0 +1,603 @@
+                // trying to write it.
+                let current = CURRENT_TASK_PTR
+                    .expect("irq before kernel started?")
+                    .as_ptr();
+
+                // Safety: we're dereferencing the current task pointer, which we're
+                // trusting the rest of this module to maintain correctly.
+                let current = usize::from((*current).descriptor().index);
+
+                let next = task::select(current, tasks);
+                let next = &mut tasks[next];
+                apply_memory_protection(next);
+                // Safety: next comes from teh task table and we don't use it again
+                // until next kernel entry, so we meet the function requirements.
+                set_current_task(next);
+            }
+
+            // Reset mtime back to 0.  In theory we could save an instruction on
+            // RV32 here and only write the low-order bits, assuming that it has
+            // been less than 12 seconds or so since our last interrupt(!), but
+            // let's avoid any possibility of a nasty surprise.
+            core::ptr::write_volatile(MTIME as *mut u64, 0);
+        })
+    }
+    crate::profiling::event_timer_isr_exit();
+}
+
+//
+// The Rust side of our trap handler after the task's registers have been
+// saved to SavedState.
+//
+#[no_mangle]
+fn trap_handler(task: &mut task::Task) {
+    match register::mcause::read().cause() {
+        //
+        // Interrupts.  Only our periodic MachineTimer interrupt via mtime is
+        // supported at present.
+        //
+        Trap::Interrupt(Interrupt::MachineTimer) => {
+            timer_handler();
+        }
+        //
+        // System Calls.
+        //
+        Trap::Exception(Exception::UserEnvCall) => {
+            unsafe {
+                // Advance program counter past ecall instruction.
+                task.save_mut().pc = register::mepc::read() as u64 + 4;
+                asm!(
+                    "
+                    la a1, CURRENT_TASK_PTR
+                    mv a0, a7               # arg0 = syscall number
+                    ld a1, (a1)             # arg1 = task ptr
+                    jal syscall_entry
+                    ",
+                );
+            }
+        }
+        //
+        // Exceptions.  Routed via the most appropriate FaultInfo.
+        //
+        Trap::Exception(Exception::IllegalInstruction) => unsafe {
+            handle_fault(task, FaultInfo::IllegalInstruction);
+        },
+        Trap::Exception(Exception::LoadFault) => unsafe {
+            handle_fault(
+                task,
+                FaultInfo::MemoryAccess {
+                    address: Some(register::mtval::read() as usize),
+                    source: FaultSource::User,
+                },
+            );
+        },
+        Trap::Exception(Exception::InstructionFault) => unsafe {
+            handle_fault(task, FaultInfo::IllegalText);
+        },
+        _ => {}
```

Comment:
This will also ignore any exceptions that aren't accounted for above. Maybe instead check if it's an exception and return only if it isn't?

Reply:
My plan is to bring back `klog` for kernel and enable some useful prints around such cases that are not handled. Currently, this panics on rv32 which doesn't really generate any log (at least when humility is not connected). So adding a panic wouldn't really help much anyways. I am tracking the activity to bring back klog and add trap handling messages in https://rivosinc.atlassian.net/browse/SW-479. Does that plan look okay?

Reply:
Having a panic would at least keep the kernel in an infinite loop, so the state of the system will be preserved, which would make it easier to trace the cause of the exception. Unless we also plan to allow the system to keep running on arbitrary exceptions? Either way, if you think it's unnecessary that's enough for me.

Reply:
Added a panic for now in a follow-up commit.